Issue #5570 Add doc for C# String.Split() (#5581)
Codecademy/docs: content/c-sharp/concepts/strings/terms/split/split.md

---
Title: '.Split()'

> Description: 'Used to break string into substrings by specifying a delimeter.'

Reviewer: ```suggestion
Description: 'Divides a string into an array of substrings based on specified delimiters.'
```

Reviewer: Description should begin with a verb

Subjects:
  - 'Computer Science'
Tags:
  - 'Strings'
  - 'Libraries'
  - 'Methods'
CatalogContent:
  - 'learn-c-sharp'
  - 'paths/computer-science'
---

> The `.Split()`** method breaks a string into substrings based on a delimeter which can be a character or another string and returns an array of strings

Reviewer: ```suggestion
The **`.Split()`** method breaks a string into substrings based on a specified delimiter, which can be a character, an array of characters, or a string. It returns an array of strings containing the split substrings.
```

Reviewer: Reframed for better clarity

## Syntax

```pseudo
// Splits a string using a character as a delimeter
string[] subs = String.Split(char)

// Splits a string using a string as a delimeter
string[] subs = String.Split(string)

// Splits a string using a character as a delimeter into maximum substrings of the number specified in the second argument.
string[] subs =  String.Split(char, Int)

// Splits a string using a character as a delimeter into maximum substrings of the number specified in the second argument.

> string[] subs =  String.Split(string, Int)

Reviewer: Can we separate these as follows:

Using a single character as a delimiter:
```pseudo
string[] subs = String.Split(char)
```

Using a string as a delimiter:
```pseudo
string[] subs = String.Split(string)
```

and others...

```

`.Split()` is a method of class `String`. This method can take upto 3 parameters:
1. The first parameter is the delimeter and it can be a `Char`, an array of characters `Char[]`, or a `String`
2. The second parameter is optional and could be the manximum number of desired substrings which is specified as an integer of type `Int32`

> 3. The third parameter which is also optional could contain the split options `StringSplitOptions`

Reviewer: The parameters should be described in the following format:
- `char`: <description>
- `string`: <description>
and other parameters

---
Title: '.Split()'

> Description: 'Divides a string into an array of substrings based on specified delimeter(s).'

Reviewer 2: There is a spelling error -

```suggestion
Description: 'Divides a string into an array of substrings based on specified delimiter(s).'
```

Subjects:

>   - 'Computer Science'

Reviewer 2: We can also add `Code Foundations` as a subject for this topic. 

```suggestion
  - 'Computer Science'
  - 'Code Foundations'
```

Tags:
  - 'Strings'
  - 'Libraries'
  - 'Methods'
CatalogContent:
  - 'learn-c-sharp'
  - 'paths/computer-science'
---

> The **`.Split()`** method breaks a string into substrings based on specified delimiter(s), which can be a character, an array of characters, or a string. It returns an array of strings containing the split substrings.

Reviewer 2: We can restructure the introduction a little to match the description - 

```suggestion
The **`.Split()`** method divides a string into an array of substrings based on specified delimiter(s), which can be a character, an array of characters, or a string.
```

## Syntax

> The various overloads of the `.Split()` method in C# are as follows:

Reviewer 2: You mention "overloads" of the .Split() method, but it might help to explicitly state that these are different ways the method can be called, depending on the type of input parameters.

```suggestion
The `.Split()` method can be used in the following ways:
```

```pseudo
Split(String[], Int32, StringSplitOptions)
```

Splits a string into a maximum number of substrings using the strings in an array as delimiters.

```pseudo
Split(Char[], Int32, StringSplitOptions)
```

Splits a string into a maximum number of substrings using the characters in an array as delimiters.

```pseudo
Split(String[], StringSplitOptions)
```

Splits a string into substrings based on the strings in an array.

```pseudo
Split(Char[])
```

Splits a string into substrings based on the characters in an array.

```pseudo
Split(Char[], StringSplitOptions)
```

Splits a string into substrings based on the characters in an array.

```pseudo
Split(Char[], Int32)
```

Splits a string into a maximum number of substrings based on the characters in an array. 

- `StringSplitOptions`: Represents an enumeration that defines whether to include empty substrings `(StringSplitOptions.None)` or exclude them `(StringSplitOptions.RemoveEmptyEntries)`.

> - `int count`: Represents the maximum number of substrings to return.

Reviewer 2: There is no usage of `int count` in above pseudo code examples, this can be confusing.

We should write the pseudo code in following way so that the datatype and the values can be more clear - 

```
Split(string[] separator, int count, StringSplitOptions options)
```

## Example

The following examples demonstrates the use of `.Split()` method:

```cs
using System;

class EqualsMethod {

    public static void Main(string[] args)
    {
        string s1 = "Rivers, Mountains, Oceans";

        string[] subs = s1.Split(',');
        foreach (var sub in subs)
        {
          Console.WriteLine($"Substring: {sub}");
        }

>        // To remove spaces, specify ', ' as the delimiter.

Reviewer 2: Here we are only removing spaces alongwith the comma, so we need to be more specific.

```suggestion
       // To remove spaces alongwith the comma, specify ', ' as the delimiter.
```

        subs = s1.Split(", ");
        foreach (var sub in subs)
        {
          Console.WriteLine($"Substring: {sub}");
        }

        // To limit the number of substrings to 2, specify the optional parameter
        subs = s1.Split(", ", 2);
        foreach (var sub in subs)
        {
          Console.WriteLine($"Substring: {sub}");
        }
    }
}
```

This example results in the following output:

```shell
Substring: Rivers
Substring:  Mountains
Substring:  Oceans
Substring: Rivers
Substring: Mountains
Substring: Oceans
Substring: Rivers
Substring: Mountains, Oceans
```

## Codebyte Example

> Run the following example to understand how the `.String()` method works:

Reviewer 2: This should be a description of what the example does. It can be something like this - 

```suggestion
This example demonstrates how the `.Split()` method works in C# for splitting a string based on delimiters (in this case, commas and spaces).
```

```codebyte/csharp
using System;

class EqualsMethod {

    public static void Main(string[] args)
    {
        string s1 = "Rivers, Mountains, Oceans";

        string[] subs = s1.Split(',');
        foreach (var sub in subs)
        {
          Console.WriteLine($"Substring: {sub}");
        }

        // To remove spaces too, we can specify ', ' as the delimeter
        subs = s1.Split(", ");
        foreach (var sub in subs)
        {
          Console.WriteLine($"Substring: {sub}");
        }
    }
}

> ```

Reviewer 2: This example is very similar to the example you have used above to explain the usage. Codebyte examples have to be a little different. You can try something like this - 

```cs
using System;

class SplitExample
{
    public static void Main(string[] args)
    {
        // A sample CSV string
        string csv = "John,25,Developer,New York,,john@example.com";

        // Split by comma (default behavior)
        string[] fields = csv.Split(',');
        Console.WriteLine("Fields after splitting by comma:");
        foreach (var field in fields)
            Console.WriteLine($"'{field}'");

        // Split by comma and remove empty entries
        string[] fieldsWithoutEmpty = csv.Split(new char[] { ',' }, StringSplitOptions.RemoveEmptyEntries);
        Console.WriteLine("\nFields after removing empty entries:");
        foreach (var field in fieldsWithoutEmpty)
            Console.WriteLine($"'{field}'");

        // Split by multiple delimiters (comma and semicolon)
        string multiDelimiterString = "apple, orange; banana,grape";
        string[] fruits = multiDelimiterString.Split(new char[] { ',', ';' });
        Console.WriteLine("\nFruits after splitting by comma and semicolon:");
        foreach (var fruit in fruits)
            Console.WriteLine($"'{fruit.Trim()}'");  // Trim spaces
    }
}
```
